File: backend/src/main/java/by/minilooth/diploma/models/bean/stores/Address.java
package by.minilooth.diploma.models.bean.stores;

import by.minilooth.diploma.models.api.AbstractEntity;
import by.minilooth.diploma.models.bean.stores.Store;
import lombok.*;

import javax.persistence.*;
import java.util.Objects;
import java.util.Set;

@Entity
@Table(name = "address")
@AllArgsConstructor
@NoArgsConstructor
@Builder
@Data
@EqualsAndHashCode(callSuper = false)
public class Address extends AbstractEntity implements Comparable<Address> {

    @Column(name = "street", nullable = false)
    private String street;

    @Column(name = "house", nullable = false)
    private String house;

    @Column(name = "housing")
    private String housing;

    @Column(name = "room")
    private String room;

    @EqualsAndHashCode.Exclude
    @ToString.Exclude
    @OneToMany(mappedBy = "address", cascade = CascadeType.ALL)
    private Set<Store> stores;

    public String getFullAddress() {
        StringBuilder sb = new StringBuilder();

        sb.append(street);
        sb.append(", д. ").append(house);

        if (!Objects.isNull(housing) && !housing.isEmpty() && !housing.isBlank()) {
            sb.append(", к. ").append(housing);
        }

        if (!Objects.isNull(room) && !room.isEmpty() && !room.isBlank()) {
            sb.append(", кв. ").append(room);
        }

        return sb.toString();
    }

    @Override
    public int compareTo(Address o) {
        return getFullAddress().compareTo(o.getFullAddress());
    }
}

File: frontend/schemas/address.ts
import * as yup from "yup";

export const AddressSchema = yup.object().shape({
  street: yup
    .string()
    .required('Это обязательное поле')
    .matches(/^[A-Za-zА-Яа-я .-]+$/, 'Улица указана в некорректном формате'),
  house: yup
    .string()
    .required('Это обязательное поле')
    .matches(/^[0-9]{1,3}[A-Za-zА-Яа-я]?$/, 'Номер дома указан в некорректном формате'),
  housing: yup
    .string()
    .matches(/^((?=[0-9])?[0-9]$|[A-Za-zА-Яа-я])$/, 'Корпус дома указан в некорректном формате'),
  room: yup
    .string()
    .matches(/^[A-Za-zА-Яа-я0-9 ]+$/, 'Номер квартиры/помещения/офиса указан в некорректном формате')
})
